designs/carbon-aware.md

```diff
@@ -0,0 +1,133 @@
+
+I'm currently working on my master's thesis in Computer Engineering (M.Sc.Eng) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (focus is AWS).
+
+RFC: https://github.com/aws/karpenter/issues/4630
+
+## Fundamentals of Green Software
+I'll try to keep it simple. The reader should be familiar with the following.
+
+A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.
+
+- **Embodied carbon emissions**: Manufacturing emissions (CO2e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance
+- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE
+
+There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).
+
+## Solution
+
+### Feature Gate
+
+The feature is proposed to be controlled using a [feature gate](https://github.com/aws/karpenter/blob/main/website/content/en/docs/concepts/settings.md#feature-gates).
+
+| **Feature**  | **Default** |         **Config Key**          | **Stage** |    **Since**    | **Until** |
+| :----------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |
+| Carbon Aware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.31.0/v0.32.0 |           |
+
+### Carbon emissions data source
+To be determined.
+
+I don't know which one should be used yet. I am currently reviewing and comparing two options:
+- [Cloud Carbon Footprint](https://github.com/cloud-carbon-footprint/ccf-coefficients). Extract data and create calculations.
+- [Boavizta](https://github.com/Boavizta/boaviztapi). Extract data and port calculations (for AWS). [Data demo](https://datavizta.boavizta.org/cloudimpact).
+
+I assume that we only want to use "static" data so we don't have to go out and make requests over the Internet to get real time data, since that would be expensive for performance.
+
+### Changes to consolidation (karpenter-core)
+
+Single Machine Consolidation (`singlemachineconsolidation.go`) and Multi Machine Consolidation (`multimachineconsolidation.go`) as well as `consolidation.go` is currently consolidating nodes to reduce costs. We want to change this when Carbon Aware is enabled. They should consolidate to minimize carbon emissions. 
+
+I have identified two potential solutions.
+
+Recommendation: solution 1.
```

Comment:
How I'd approach this:
- set a price per tonne (or kg) of CO₂e
  - maybe have a config option for this, defaulting to $0. A good starting value for taking environmental impact into account is $0.25 / kg
- adjust the price algorithm to account for the pollution cost supplement

That could be consolidation option 3.

Reply:
Also, the same pricing choices should influence instance launch for reasons other than consolidation.

Reply:
@sftim That is actually a very good idea!! Have not thought of that before. That way one could "simulate" a CO₂e tax/price on the machines.

Reply:
@sftim I have added your idea as option 3. Interesting approach, I like it